RDI-2909: add latency measurements based on probes and system clock

diegonieto opened the pull request:
No network supported

diegonieto committed:
video-encoder-stats: create vmaf property and handle optional behaviour

* Detect in runtime whether vmaf element is available
* Check whether property is enabled. Default: enabled
* Expose decode_src pad only if vmaf pipeline branch is enabled
* Add a new example not using vmaf

diegonieto committed:
video-encoder-stats: handle input queues internally

diegonieto committed:
video-encoder-stats: add latency metric

* get latency from SystemTime
** TODO: implement npt as webrtc-precise-sync-send.rs does
* add timestamp before encoding as part of videoencoderstatsmeta
* compare the input timestamp in video-compare-mixer with the current
timestamp (after decoding the video stream)
* add a new example video-stats-zero-latency to showcase the latency
difference

rgonzalezfluendo commented on video/stats/src/encoderstats/imp.rs:
```diff
@@ -185,6 +150,97 @@ impl EncoderStats {
 
         encoder.set_property("name", "enc");
 
+        // Add internal queue at the beginning
```
Why is needed?

diegonieto replied:
This is because we use the queue thread to get the CPU time of the encoder. Since the encoder is wrapped in a unique thread distinguished by the queue name, we can get that encoder CPU time. If the encoder spawns more specific threads, like `lcevch` does, then we need also to add them. Does it make sense this approach?

rgonzalezfluendo commented on video/stats/src/encoderstats/imp.rs:
```diff
@@ -185,6 +150,97 @@ impl EncoderStats {
+            };
+            let (mut total_utime, mut total_stime) = get_cpu_usage(thread_name.to_string());
+
+            if stats.name == "flulcevch264enc" {
+                // flulcevc uses multiple threads, so we need to get the CPU usage of all threads
+                let (utime, stime) = get_cpu_usage("lcevc".to_string());
+                gst::log!(CAT, "flulcevc lcevc utime: {}, stime: {}", utime, stime);
+                // Add the  CPU usage to the total CPU usage
+                total_utime += utime;
+                total_stime += stime;
+
+                let (utime, stime) = get_cpu_usage("pool.".to_string());
+                gst::log!(CAT, "flulcevc pool utime: {}, stime: {}", utime, stime);
+                // Add the  CPU usage to the total CPU usage
+                total_utime += utime;
+                total_stime += stime;
+            }
+
+            if stats.name == "lcevch264enc" {
+                // lcevc uses multiple threads, so we need to get the CPU usage of all threads
+                let (utime, stime) = get_cpu_usage("pool.".to_string());
+                gst::log!(CAT, "lcevc pool utime: {}, stime: {}", utime, stime);
+                // Add the  CPU usage to the total CPU usage
+                total_utime += utime;
+                total_stime += stime;
+            }
+
```
How ugly to have to implement this.

I thought the name of the pthreads was imposed by GStreamer, even for third-party ones. Maybe I thought too much.

We need to do it more generic to be able to upstream.

diegonieto replied:
At least for `lcevc` there are some specific threads spawned by the backend that always match that prefix

> We need to do it more generic to be able to upstream.
I think the name is given by the one that spawn the thread. So, if the encoder spawns X threads underneath, not sure whether we can modify their name.

We can use a `match` clause that maybe is more readable. I will double-check

diegonieto replied:
I have checking different set-up:
| x264enc threads | encq0:src instances |
|-----------------|-------------------|
| 1 | 1 |
| 2 | 8-10 |
| 4 | 18 |
| 8 | 30-34 |
| 16 | 50 |

If not set, it reaches 36 threads, as setting it to 8, the number of threads of my CPU.

[x264enc-threads-4.log](https://github.com/user-attachments/files/22335893/x264enc-threads-4.log)
[x264enc-threads-1.log](https://github.com/user-attachments/files/22335894/x264enc-threads-1.log)
Pipeline:
```
souphttpsrc location=\"https://ftp.nluug.nl/pub/graphics/blender/demo/movies/ToS/tears_of_steel_1080p.mov\" ! qtdemux name=demux demux.video_0 ! queue ! decodebin3 ! videoconvertscale ! capsfilter caps=\"video/x-raw,aspect-ratio=1/1\" ! tee name=tee ! video-encoder-stats encoder=\"x264enc bitrate=1024 tune=zerolatency speed-preset=ultrafast threads=4\" name=vs0 vmaf-stats=false tee. ! video-encoder-stats encoder=\"x264enc bitrate=512 tune=zerolatency speed-preset=ultrafast\" name=vs1 video-compare-mixer split-screen=true backend=CPU name=mixer vs0.src ! h264parse ! avdec_h264 ! mixer.sink_0  vs1.src ! h264parse ! avdec_h264 ! mixer.sink_1  mixer. ! autovideosink
```

rgonzalezfluendo commented on video/stats/src/encoderstats/imp.rs:
```diff
@@ -132,6 +80,9 @@ impl EncoderStats {
             };
             stats.lock().unwrap().buffer_in();
             gst::log!(CAT, "Buffer in encoder sink pad");
+            stats.lock().unwrap().pre_encode_time = *gst::ClockTime::from_nseconds(
```
Note encoders have internal queues. Reorder buffers inside the encoder is normal. We can not assume one buffer in, one buffer out.

For more info check discussion [here](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/9622)

Better solution should be to create the meta before buffer goes inside the encoder and to set `pre_encode_time` in the meta, not in a global stats.

diegonieto replied:
Totally, this is now reporting well.

While `input_queue_src_pad.add_probe()` accomplish
> Better solution should be to create the meta before buffer goes inside the encoder and to set pre_encode_time in the meta, not in a global stats.

by creating the meta before the buffer goes to the encoder, `encoder_sink_pad.add_probe()` & `encoder_src_pad.add_probe()` update global variable affecting the next upcoming buffer, instead of updating the meta available in the current one. This is wrong. Task created here: https://fluendo.atlassian.net/browse/RDI-2963

diegonieto committed:
video-encoder-stats: let vmaf score a N/A by default

If vmaf metric is not available N/A will be shown in the metric instead of 0 value.

diegonieto committed:
video-encoder-stats: add silent property and post stats as messages

This enables posting the stats as messages unless the new property silent is set true, by default is false

diegonieto commented:
This PR will be closed, all these commits are now part of https://github.com/fluendo/gst-plugins-rs/pull/4

rgonzalezfluendo commented on video/stats/src/encoderstats/imp.rs:
```diff
@@ -554,6 +582,15 @@ impl ObjectImpl for EncoderStats {
+                glib::ParamSpecBoolean::builder("silent")
+                    .nick("Silent")
+                    .blurb("Enable silent mode (disable stdout output of stats)")
+                    .default_value(false)
```
`.default_value(true)`
